server/aesdsocket.c

```diff
@@ -0,0 +1,287 @@
+	int send_buf_size = 0;
+
+	while (1) {
+        // accept connection from client
+        client_fd = accept(socket_num, (struct sockaddr*) &client_sockaddr, &client_addrlen);
+        if (client_fd == -1) {
+            perror("accept");
+            return -1;
+        }
+
+		// log IP to syslog
+        char* ip_addr = inet_ntoa(client_sockaddr.sin_addr);
+        syslog(LOG_DEBUG, "Accepted connection from %s\n", ip_addr);
+
+		int recv_buf_pos = 0;
+		long recv_buf_size = HUNDRED_BYTES;
+		
+		char* recv_buf = malloc(recv_buf_size * sizeof(char));
+        if (recv_buf == NULL) {
+            perror("malloc failure");
+            return -1;
+        }
+
+        // mask signals
+        status = sigprocmask(SIG_BLOCK, &cur_set, &prev_set);
+        if (status == -1) {
+            printf("signal masking failed\n");
+        }
+
+        // receive data from client
+        // each loop iteration is 1 newline of input
+        while (1) {
+            // save up to 100 bytes into buf
+			num_bytes = recv(client_fd, buf, HUNDRED_BYTES, 0);
+
+            // exit loop on error
+            if (num_bytes == -1) {
+                perror("recv");
+                return -1;
+            }
+
+			// check if allocated buf size is sufficient
+			if(recv_buf_pos + num_bytes > recv_buf_size) {
+				recv_buf_size += num_bytes;
+				recv_buf = realloc(recv_buf, recv_buf_size * sizeof(char));
+			}
+			
+			// copy buf to recv_buf
+			memcpy(recv_buf + recv_buf_pos, buf, num_bytes);
+			recv_buf_pos += num_bytes;
+
+            // exit loop if there was a new line character received
+			if(strchr(buf, '\n') != NULL)
+				break;
+        }
+
+        // unmask signals
+        status = sigprocmask(SIG_UNBLOCK, &prev_set, NULL);
+        if (status == -1) {
+            printf("signal unmasking failed\n");
+        }
+		
+		// write new bytes to file
+		num_bytes = write(file_fd, recv_buf, recv_buf_pos);
+        printf("** Writing %d bytes\n", num_bytes);
+		if (num_bytes == -1 || num_bytes != recv_buf_pos) {
+			perror("write");
+			return -1;
+		}
+		
+		// go to start of file
+		lseek(file_fd, 0, SEEK_SET);
+		
+		// allocate memory for send buffer
+		send_buf_size += recv_buf_pos;
+		char* send_buf = malloc(send_buf_size * sizeof(char));
+        if (send_buf == NULL) {
+            perror("malloc failure");
+            return -1;
+        }
+			
+		// read the ENTIRE file
+		num_bytes = read(file_fd, send_buf, send_buf_size);
+        printf("** Reading %d bytes\n", send_buf_size);
+		if (num_bytes == -1 || num_bytes != send_buf_size) {
+			perror("read");
+			return -1;
+		}
+
+        // mask signals
+        status = sigprocmask(SIG_BLOCK, &cur_set, &prev_set);
+        if (status == -1) {
+            printf("signal masking failed\n");
+        }
+		
+		// send data to the client
+		num_bytes = send(client_fd, send_buf, num_bytes, 0);
+		if (num_bytes == -1 || num_bytes != send_buf_size) {
+			perror("send");
+			return -1;
```

Comment:
@bjornhnelson missing cleanup in error paths, consider a design where you always fall through in success or failure cases (or after signal) and always do the same cleanup.